src/org/traccar/FilterHandler.java

```diff
@@ -126,33 +132,35 @@ private boolean filterDistance(Position position, Position last) {
     private boolean filterMaxSpeed(Position position, Position last) {
         if (filterMaxSpeed != 0 && last != null) {
             double distance = position.getDouble(Position.KEY_DISTANCE);
-            long time = position.getFixTime().getTime() - last.getFixTime().getTime();
+            double time = position.getFixTime().getTime() - last.getFixTime().getTime();
             return UnitsConverter.knotsFromMps(distance / (time / 1000)) > filterMaxSpeed;
         }
         return false;
     }
 
     private boolean filterLimit(Position position, Position last) {
-        if (filterLimit != 0) {
-            if (last != null) {
-                return (position.getFixTime().getTime() - last.getFixTime().getTime()) > filterLimit;
-            } else {
-                return false;
-            }
-        } else {
-            return false;
+        if (filterLimit != 0 && last != null) {
+            return (position.getFixTime().getTime() - last.getFixTime().getTime()) <= filterLimit;
         }
+        return false;
     }
 
     private boolean filter(Position position) {
 
+        if (keepAlarms && position.getAttributes().containsKey(Position.KEY_ALARM)) {
+            return false;
+        }
+
         StringBuilder filterType = new StringBuilder();
 
         Position last = null;
         if (Context.getIdentityManager() != null) {
             last = Context.getIdentityManager().getLastPosition(position.getDeviceId());
         }
 
+        if (filterLimit(position, last)) {
```

Comment:
I think you misunderstood how limit works. It's works similar to "keepAlarms". Basically, if time between last position is more than X, then don't filter. Maybe we should rename it to avoid confusion.

Reply:
I have inverted the comparison, now it is like other functions

Reply:
I see that, but limit is not for filtering, it's for preventing any other filtering.

Reply:
Got it